### PR TITLE
Fix #212: Detailed required vs. not-required assessment

### LINKED ISSUE #212: need to perform a v1 detailed required vs. not-required assessment

### PR DESCRIPTION
**Pull request by ghost**
Also:
o Make a couple of dozen other copy edits.

### CODE REVIEW

File: document/FileFormat.mdk [hunk 1]
```diff
@@ -1908,7 +1915,7 @@ namespace A {
 
 (These could not coexist in the same compilation, but there is no reason two such source files could not exist.)
 
-If the tool detected one result in class `B` in file1.cpp, and another result in namespace `B` in file2.cpp,
+If the tool detected one result in class `B` in _file1.cpp_, and another result in namespace `B` in _file2.cpp_,
 the `fullyQualifiedLogicalName` for both would be `A::B`.
 In that case, the tool might set the `logicalLocationKey` property in either one of the results to `A::B-0`,
```
Comment (michaelcfanning) [inline document/FileFormat.mdk]: Make this A::B-1, as in your rule key example above. Presumably, A::B already exists, so we assume that's the 0th item. :) #Resolved

File: document/FileFormat.mdk [hunk 2]
```diff
@@ -2385,7 +2390,7 @@ containing a message relevant to this call stack.
 
 A `stack` object shall contain a property named `frames` whose value is an array of one or more `stackFrame` objects (&sect;[#stackFrame]).
 This array shall include every function call in the stack for which the tool has information,
-and the entries that are present shall occur in chronological order with the most recent (innermost) call first and the least recent (outermost) call last.
+and the entries that are present shall occur in nesting order with the most recent (innermost) call first and the least recent (outermost) call last.
```
Comment (michaelcfanning) [inline document/FileFormat.mdk]: I don't understand this change and it doesn't clarify anything. what is 'nesting' order? previously, the statement was clear #Resolved

File: document/FileFormat.mdk [hunk 3]
```diff
@@ -2857,9 +2865,28 @@ The value of `formatId` shall correspond to one of the names in the set of name/
 of the `rule` object (&sect;[#rule]) whose `id` property (&sect;[#rule-id]) matches the `ruleId` property (&sect;[#result-ruleId]) of this result.
 
 ### `arguments` property { #formattedMessage-arguments }
-A `formattedMessage` object may contain a property named `arguments`,
+If the message format string specified by `formatId` contains any placeholders,
+the `formattedMessage` object shall contain a property named `arguments`,
 whose value is an array of string values that will be used, in combination with a message format,
 to construct a result message.
+The array shall have as many elements are there are distinct placeholders in the message format.
```
Comment (michaelcfanning) [inline document/FileFormat.mdk:2872]: it's clear in the example but do we mention elsewhere explicitly that these are 0-indexed? #ByDesign

**Comment by michaelcfanning** [comment]
+1 modulo 'nesting' change #Closed

File: document/FileFormat.mdk (hunk 1 above)
Comment (ghost) [inline document/FileFormat.mdk]: Fixed.

---

In reply to: [72723560](https://github.com/sarif-standard/sarif-spec/pull/255#discussion_r72723560) [](ancestors = 72723560)

File: document/FileFormat.mdk (hunk 3 above)
Comment (ghost) [inline document/FileFormat.mdk:2872]: Yes, we do: "Each placeholder shall be of the form "`{`_n_`}`", where _n_ is a non-negative integer which represents a 0-based index into the list of arguments."

---

In reply to: [72723920](https://github.com/sarif-standard/sarif-spec/pull/255#discussion_r72723920) [](ancestors = 72723920)

File: document/FileFormat.mdk (hunk 2 above)
Comment (ghost) [inline document/FileFormat.mdk]: Meh ;-) I really didn't like "chronological" order when I re-read this, and I thought "nesting" was clear, but I'm not adamant about it. I'll change it back.

---

In reply to: [72723785](https://github.com/sarif-standard/sarif-spec/pull/255#discussion_r72723785) [](ancestors = 72723785)

**Comment by ghost** [comment]
Merged

---

In reply to: [236059506](https://github.com/sarif-standard/sarif-spec/pull/255#issuecomment-236059506) [](ancestors = 236059506)

**Pull request merged**